ext/contributioncancelactions/contributioncancelactions.php

```diff
@@ -4,16 +4,40 @@
 // phpcs:disable
 use CRM_Contributioncancelactions_ExtensionUtil as E;
 // phpcs:enable
+use Civi\Api4\LineItem;
 
 /**
  * Implements hook_civicrm_preProcess().
  *
+ * This enacts the following
+ * - find and cancel any related pending memberships
+ * - (not yet implemented) find and cancel any related pending participant records
+ * - (not yet implemented) find any related pledge payment records. Remove the contribution id.
+ *
  * @link https://docs.civicrm.org/dev/en/latest/hooks/hook_civicrm_post
  */
 function contributioncancelactions_civicrm_post($op, $objectName, $objectId, $objectRef) {
   if ($op === 'edit' && $objectName === 'Contribution') {
-    if ($objectRef->contribution_status_id === CRM_Core_PseudoConstant::getKey('CRM_Contribute_BAO_Contribution', 'contribution_status_id', 'Cancelled')) {
-      // Do the stuff.
+    if ('Cancelled' === CRM_Core_PseudoConstant::getName('CRM_Contribute_BAO_Contribution', 'contribution_status_id', $objectRef->contribution_status_id)) {
```

Comment:
@eileenmcnaughton The issue here is that it will trigger *every* time a cancelled contribution is edited. It may not be common to edit one but it's certainly possible. And there are various instances when eg. contribution.create gets called multiple times for different reasons.
One example being an extension like giftaid, if you run an update to recalculate amounts it would trigger a contribution.create.
The main issue with being triggered multiple times is that any notifications/logic tied to the memberships may get triggered and for example, a contact might receive a "you've cancelled your membership" notification weeks afterwards. That can normally be avoided in eg. CiviRules by specifying a condition of "Membership status changed from Pending to Cancelled".

Reply:
@mattwire so the protection here is the check on what the membership status is  - it won't update the membership if the status is not Pending. 

The current core logic is really that Pending memberships should be cancelled if the Contribution is cancelled - so in theory this should do nothing on already-cancelled-contributions (except for an extra query but this isn't a high traffic flow).